File: web/src/lib/adjust.ts
/* Corporate-action adjustment, applied when data is READ.
 *
 * A direct port of `backend/btb/data/adjust.py`. The two implementations must
 * agree: if the chart shows a different split-adjusted price than the engine
 * backtests on, one of them is lying and there is no way to tell which from
 * looking at either.
 *
 * `bars` holds raw prints. Every adjusted view is derived here from the events
 * table, which is what keeps stored rows permanent -- when a symbol splits next
 * year, nothing on disk changes meaning.
 *
 *   "none"   raw prints -- what traded, at the price it traded
 *   "split"  split-adjusted. The default for charts and backtests.
 *   "total"  splits and dividends, i.e. total return. For benchmarks.
 *
 * Splits are applied before dividends, because the venue quotes dividend
 * amounts already split-adjusted; applying one against a raw close subtracts a
 * nonsense fraction.
 */

export type Bar = {
  ts: number;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
  partial: boolean;
};

export type MarketEvent = {
  ts: number;
  kind: 'split' | 'dividend';
  ratio: number | null;
  amount: number | null;
};

export type AdjustMode = 'none' | 'split' | 'total';

const DAY = 86400;

/* Which bars fall strictly before a corporate action.
 *
 * An ex-date is stamped at the session open (13:30 or 14:30 UTC) while daily
 * bars are stored snapped to UTC midnight. Comparing those directly puts the
 * ex-date bar before its own event and adjusts one bar too many -- the ex-date
 * bar is already trading post-split. So daily-or-slower series compare at day
 * granularity; intraday series compare raw timestamps, where the session-open
 * stamp already separates the ex-date's bars correctly.
 */
function isPrior(barTs: number, eventTs: number, step: number): boolean {
  if (step >= DAY) return Math.floor(barTs / DAY) < Math.floor(eventTs / DAY);
  return barTs < eventTs;
}

export function tfSeconds(timeframe: string): number {
  const units: Record<string, number> = { m: 60, h: 3600, d: 86400, w: 604800 };
  const unit = timeframe.slice(-1);
  const n = Number(timeframe.slice(0, -1));
  if (!units[unit] || !Number.isFinite(n) || n <= 0) {
    throw new Error(`unrecognised timeframe ${timeframe}`);
  }
  return n * units[unit];
}

export function adjustBars(
  bars: Bar[],
  events: MarketEvent[],
  mode: AdjustMode,
  timeframe: string,
): Bar[] {
  if (mode === 'none' || bars.length === 0 || events.length === 0) return bars;

  const step = tfSeconds(timeframe);
  const priceFactor = new Float64Array(bars.length).fill(1);
  const volumeFactor = new Float64Array(bars.length).fill(1);

  // --- splits, newest first so the factors compound --------------------------
  const splits = events
    .filter((e) => e.kind === 'split' && (e.ratio ?? 0) > 0)
    .sort((a, b) => b.ts - a.ts);

  for (const s of splits) {
    const ratio = s.ratio as number;
    for (let i = 0; i < bars.length; i++) {
      if (isPrior(bars[i].ts, s.ts, step)) {
        priceFactor[i] *= 1 / ratio;
        volumeFactor[i] *= ratio;
      }
    }
  }

  if (mode === 'total') {
    // Dividends are computed against SPLIT-ADJUSTED closes, since that is the
    // basis the venue quotes the amounts on.
    const splitClose = bars.map((b, i) => b.close * priceFactor[i]);
    const dividends = events
      .filter((e) => e.kind === 'dividend' && (e.amount ?? 0) > 0)
      .sort((a, b) => b.ts - a.ts);

    for (const d of dividends) {
      const firstOnOrAfter = bars.findIndex((b) => !isPrior(b.ts, d.ts, step));
      if (firstOnOrAfter <= 0) continue; // event outside the window we hold
      const c = splitClose[firstOnOrAfter];
      const amt = d.amount as number;
      // A dividend at or above the close drives the factor to zero or negative.
      // That is bad event data, not a corporate action, and must never be
      // allowed to invert a price series.
      if (!(c > 0) || amt >= c) continue;
      const f = 1 - amt / c;
      for (let i = 0; i < bars.length; i++) {
        if (isPrior(bars[i].ts, d.ts, step)) priceFactor[i] *= f;
      }
    }
  }

  return bars.map((b, i) => ({
    ...b,
    open: b.open * priceFactor[i],
    high: b.high * priceFactor[i],
    low: b.low * priceFactor[i],
    close: b.close * priceFactor[i],
    volume: b.volume * volumeFactor[i],
  }));
}

File: backend/btb/data/adjust.py
"""Corporate-action adjustment, applied when data is READ, never when stored.

`bars` holds what actually printed. This module turns that into whichever view
a caller needs. Doing it here rather than at ingest is what makes the stored
rows permanent facts: when a symbol splits next year, nothing on disk changes
meaning, because nothing on disk was ever adjusted.

Three views:

    "none"    raw prints -- what traded, at the price it traded.
    "split"   split-adjusted. The default for backtesting.
    "total"   splits and dividends, i.e. total return. For benchmarks.

Split-adjusted is the backtest default because a split changes what a price
means while a dividend does not. A 2:1 split makes last year's 108 dollars and
today's 54 the same thing. Re-basing every past price for a dividend, by
contrast, invents prices at which nothing ever traded -- correct for measuring
total return, wrong for simulating trades.

Whichever view produced a number must be recorded with it. A strategy compared
across two adjustment modes is being compared across two different assets.

**Order matters.** Splits are applied before dividends, because Yahoo reports
dividend amounts already split-adjusted (AAPL's 2012 dividend is quoted as
0.0946, the $2.65 actually paid divided by the 28x of later splits). Applying a
split-adjusted dividend against a raw close would subtract a nonsense fraction.
"""
from __future__ import annotations

import numpy as np
import pandas as pd

from .bars import tf_seconds

MODES = ("none", "split", "total")
DAY = 86400


def _prior_mask(bar_ts: np.ndarray, event_ts: int, step: int) -> np.ndarray:
    """Which bars fall strictly before a corporate action.

    The subtlety that has bitten twice: an ex-date is stamped at the SESSION
    OPEN (13:30 or 14:30 UTC), while daily bars are stored snapped to UTC
    midnight. Comparing those directly puts the ex-date bar at 00:00 before its
    own 13:30 event, and adjusts one bar too many -- the ex-date bar is already
    trading post-split.

    So daily-or-slower series compare at DAY granularity; intraday series
    compare raw timestamps, where the session-open stamp already separates the
    ex-date's bars correctly.
    """
    if step >= DAY:
        return (bar_ts // DAY) < (event_ts // DAY)
    return bar_ts < event_ts


def adjust(df: pd.DataFrame, events: dict, mode: str, timeframe: str) -> pd.DataFrame:
    """Convert RAW bars to the requested view.

    Back-adjustment convention: the most recent bar keeps its actual price and
    history is scaled to meet it. That is the right way round for a backtest --
    today's price is the one a user recognises, and forward-adjusting would
    change today's number every time a future event occurred.

    Per event at time t, every bar strictly before t is scaled:

        split of ratio R (7:1 -> R=7):     prices x 1/R, volume x R
        dividend of D on a close of C:     prices x (1 - D/C)

    Factors accumulate multiplicatively, so successive splits and a long
    dividend history compound correctly.
    """
    if mode not in MODES:
        raise ValueError(f"unknown adjustment mode {mode!r}; expected {MODES}")
    if df.empty or mode == "none":
        return df

    step = tf_seconds(timeframe)
    out = df.copy()
    ts = out["ts"].to_numpy(dtype="int64")
    pf = np.ones(len(out), dtype="float64")
    vf = np.ones(len(out), dtype="float64")

    # --- splits, newest first so the factors compound ------------------------
    for s in sorted(events.get("splits", []), key=lambda x: x["ts"], reverse=True):
        ratio = float(s.get("ratio") or 0)
        if ratio <= 0:
            continue
        prior = _prior_mask(ts, int(s["ts"]), step)
        if not prior.any():
            continue
        pf[prior] *= 1.0 / ratio
        vf[prior] *= ratio

    if mode == "total":
        # Dividends are computed against SPLIT-ADJUSTED closes, since that is
        # the basis the venue quotes the amounts on.
        split_close = out["close"].to_numpy(dtype="float64") * pf
        for d in sorted(events.get("dividends", []), key=lambda x: x["ts"], reverse=True):
            amt = float(d.get("amount") or 0)
            if amt <= 0:
                continue
            t = int(d["ts"])
            prior = _prior_mask(ts, t, step)
            on_after = np.flatnonzero(~prior)
            if not prior.any() or len(on_after) == 0:
                continue                    # event outside the window we hold
            c = float(split_close[on_after[0]])
            # A dividend at or above the close would drive the factor to zero or
            # negative. That is bad event data, not a real corporate action, and
            # it must never be allowed to invert a price series.
            if c <= 0 or amt >= c:
                continue
            pf[prior] *= (1.0 - amt / c)

    for col in ("open", "high", "low", "close"):
        out[col] = out[col].to_numpy(dtype="float64") * pf
    out["volume"] = out["volume"].to_numpy(dtype="float64") * vf
    return out


def unadjust_splits(df: pd.DataFrame, events: dict, timeframe: str) -> pd.DataFrame:
    """Recover raw prints from a feed that ships split-adjusted prices.

    Yahoo's OHLC arrives already split-adjusted while still carrying the event
    stream, so storing it verbatim would store an adjusted price under a column
    that promises a raw one -- and the error is invisible until the next split
    makes those numbers mean something new.

    This is the inverse of `adjust(..., "split")`, applied once at ingest.
    """
    if df.empty:
        return df

    step = tf_seconds(timeframe)
    out = df.copy()
    ts = out["ts"].to_numpy(dtype="int64")
    pf = np.ones(len(out), dtype="float64")
    vf = np.ones(len(out), dtype="float64")

    for s in events.get("splits", []):
        ratio = float(s.get("ratio") or 0)
        if ratio <= 0:
            continue
        prior = _prior_mask(ts, int(s["ts"]), step)
        if not prior.any():
            continue
        pf[prior] *= ratio
        vf[prior] *= 1.0 / ratio

    for col in ("open", "high", "low", "close"):
        out[col] = out[col].to_numpy(dtype="float64") * pf
    out["volume"] = out["volume"].to_numpy(dtype="float64") * vf
    return out
